### PR TITLE
Optimizations for scala.collection.{immutable,mutable}.BitSet#diff

### PR DESCRIPTION
**Pull request by joshlemer**
Summary of changes:

- `immutable.BitSet.empty` is now a `final val` rather than a `def`

- `immutable.{BitSet1, BitSet2}` have hand-crafted implementations of `diff` 

- `immutable.BitSetN` and `mutable.BitSet`s implementations of `diff` has been optimized to reduce allocation through two optimizations:
  - the inner `Array[Long]`s are traversed in reverse, and the method takes note of the first result word which is non-zero. The index of this word will be the maximum index in the resulting BitSet, and and so an Array of exactly the right size can be allocated once. This avoids allocating too big of an array, and also avoids building a new BitSet from empty, which would repeatedly double its array until it has enough capacity.
  - in the case of `immutable.BitSetN`, the implementation of `diff` will return `this` if there are no changes made in the Set. It does this by tracking whether or not any result word has differed from its original word, and it avoids allocating an array entirely if it gets all the way to the beginning of the BitSet without seeing any changes. 

- `mutable.BitSet`'s method `&~= (BitSet): BitSet` no longer ensures it is as big as its argument, no need to since we're only removing elements. Also the method's use of `Range` and `for` has been replaced with `while`

- both `mutable.BitSet` and `immutable.BitSet{1,2,N}` have new and optimized implementations of filter which perform similar optimizations. Benchmarks below!

## Benchmarks
Here are some benchmarks. The way to interpret them are:
* method under test: `left.diff(right)`
* percentageFull: the approximate percent of bits in both `left` and `right` which are set to `1`
* leftSize: the approximate max int contained in the `left` BitSet (may be significantly lower for the case of percentageFull = 0.01, since many Int's are not present)
* rightSize: same as leftSize, but for right.

Note I only did some quick benchmarks (1 fork, 6 warmup, 6 normal iterations) because there are a lot of combinations.

### Summary of results

Most configurations of parameters yield an improvement of ~3-10x

```

[info] Benchmark                (percentageFull)  (sizeLeft)  (sizeRight)  Mode  Cnt        Score         Error  Units
[info] BitSetBenchmark.newDiff              0.01           1            1  avgt    6       31.208 ±       8.739  ns/op
[info] BitSetBenchmark.newDiff              0.01           1         1000  avgt    6       45.754 ±      52.256  ns/op
[info] BitSetBenchmark.newDiff              0.01           1     10000000  avgt    6       35.929 ±       4.026  ns/op
[info] BitSetBenchmark.newDiff              0.01        1000            1  avgt    6       57.548 ±      38.719  ns/op
[info] BitSetBenchmark.newDiff              0.01        1000         1000  avgt    6      246.341 ±     556.715  ns/op
[info] BitSetBenchmark.newDiff              0.01        1000     10000000  avgt    6      219.376 ±     325.516  ns/op
[info] BitSetBenchmark.newDiff              0.01    10000000            1  avgt    6       52.386 ±      21.380  ns/op
[info] BitSetBenchmark.newDiff              0.01    10000000         1000  avgt    6      163.745 ±      49.906  ns/op
[info] BitSetBenchmark.newDiff              0.01    10000000     10000000  avgt    6  4097497.984 ±  737797.615  ns/op
[info] BitSetBenchmark.newDiff               0.3           1            1  avgt    6       36.079 ±      30.826  ns/op
[info] BitSetBenchmark.newDiff               0.3           1         1000  avgt    6       44.654 ±      13.963  ns/op
[info] BitSetBenchmark.newDiff               0.3           1     10000000  avgt    6       40.285 ±      12.105  ns/op
[info] BitSetBenchmark.newDiff               0.3        1000            1  avgt    6       98.747 ±     246.763  ns/op
[info] BitSetBenchmark.newDiff               0.3        1000         1000  avgt    6      280.314 ±       9.058  ns/op
[info] BitSetBenchmark.newDiff               0.3        1000     10000000  avgt    6      294.016 ±      72.456  ns/op
[info] BitSetBenchmark.newDiff               0.3    10000000            1  avgt    6   682123.637 ± 2963731.741  ns/op
[info] BitSetBenchmark.newDiff               0.3    10000000         1000  avgt    6  1979937.589 ±  130511.341  ns/op
[info] BitSetBenchmark.newDiff               0.3    10000000     10000000  avgt    6  3340775.918 ±  621110.908  ns/op
[info] BitSetBenchmark.newDiff               0.9           1            1  avgt    6       40.548 ±      38.187  ns/op
[info] BitSetBenchmark.newDiff               0.9           1         1000  avgt    6       40.362 ±      12.110  ns/op
[info] BitSetBenchmark.newDiff               0.9           1     10000000  avgt    6       36.276 ±      20.626  ns/op
[info] BitSetBenchmark.newDiff               0.9        1000            1  avgt    6      167.517 ±       1.397  ns/op
[info] BitSetBenchmark.newDiff               0.9        1000         1000  avgt    6      279.044 ±       6.620  ns/op
[info] BitSetBenchmark.newDiff               0.9        1000     10000000  avgt    6      284.226 ±      19.683  ns/op
[info] BitSetBenchmark.newDiff               0.9    10000000            1  avgt    6  2067804.998 ±  216944.640  ns/op
[info] BitSetBenchmark.newDiff               0.9    10000000         1000  avgt    6  1933810.397 ±  303777.991  ns/op
[info] BitSetBenchmark.newDiff               0.9    10000000     10000000  avgt    6  4975641.773 ± 4786001.431  ns/op
[info] BitSetBenchmark.newDiff               1.0           1            1  avgt    6       28.446 ±       4.065  ns/op
[info] BitSetBenchmark.newDiff               1.0           1         1000  avgt    6       38.800 ±      26.688  ns/op
[info] BitSetBenchmark.newDiff               1.0           1     10000000  avgt    6       34.487 ±       1.586  ns/op
[info] BitSetBenchmark.newDiff               1.0        1000            1  avgt    6      169.287 ±       3.249  ns/op
[info] BitSetBenchmark.newDiff               1.0        1000         1000  avgt    6      145.236 ±       0.743  ns/op
[info] BitSetBenchmark.newDiff               1.0        1000     10000000  avgt    6      146.821 ±       9.846  ns/op
[info] BitSetBenchmark.newDiff               1.0    10000000            1  avgt    6  2132533.344 ±  318796.881  ns/op
[info] BitSetBenchmark.newDiff               1.0    10000000         1000  avgt    6  2209938.828 ± 1484112.686  ns/op
[info] BitSetBenchmark.newDiff               1.0    10000000     10000000  avgt    6  2777661.491 ± 2940784.395  ns/op
[info] BitSetBenchmark.oldDiff              0.01           1            1  avgt    6      737.497 ±     856.767  ns/op
[info] BitSetBenchmark.oldDiff              0.01           1         1000  avgt    6      470.978 ±     148.842  ns/op
[info] BitSetBenchmark.oldDiff              0.01           1     10000000  avgt    6      431.590 ±       4.583  ns/op
[info] BitSetBenchmark.oldDiff              0.01        1000            1  avgt    6      694.960 ±      92.029  ns/op
[info] BitSetBenchmark.oldDiff              0.01        1000         1000  avgt    6      748.125 ±      12.566  ns/op
[info] BitSetBenchmark.oldDiff              0.01        1000     10000000  avgt    6      740.510 ±      10.169  ns/op
[info] BitSetBenchmark.oldDiff              0.01    10000000            1  avgt    6  5826771.627 ± 9064974.433  ns/op
[info] BitSetBenchmark.oldDiff              0.01    10000000         1000  avgt    6  6146382.193 ± 3875187.812  ns/op
[info] BitSetBenchmark.oldDiff              0.01    10000000     10000000  avgt    6  4580814.501 ± 2271609.579  ns/op
[info] BitSetBenchmark.oldDiff               0.3           1            1  avgt    6      423.621 ±       5.445  ns/op
[info] BitSetBenchmark.oldDiff               0.3           1         1000  avgt    6      433.577 ±       9.308  ns/op
[info] BitSetBenchmark.oldDiff               0.3           1     10000000  avgt    6      438.545 ±      31.578  ns/op
[info] BitSetBenchmark.oldDiff               0.3        1000            1  avgt    6      705.773 ±       9.507  ns/op
[info] BitSetBenchmark.oldDiff               0.3        1000         1000  avgt    6      743.735 ±      19.361  ns/op
[info] BitSetBenchmark.oldDiff               0.3        1000     10000000  avgt    6      764.688 ±      65.624  ns/op
[info] BitSetBenchmark.oldDiff               0.3    10000000            1  avgt    6  4417984.005 ± 1421879.462  ns/op
[info] BitSetBenchmark.oldDiff               0.3    10000000         1000  avgt    6  5127837.605 ±  834225.922  ns/op
[info] BitSetBenchmark.oldDiff               0.3    10000000     10000000  avgt    6  4711617.779 ± 2122533.210  ns/op
[info] BitSetBenchmark.oldDiff               0.9           1            1  avgt    6      426.939 ±       4.010  ns/op
[info] BitSetBenchmark.oldDiff               0.9           1         1000  avgt    6      432.011 ±       7.210  ns/op
[info] BitSetBenchmark.oldDiff               0.9           1     10000000  avgt    6      449.853 ±      69.014  ns/op
[info] BitSetBenchmark.oldDiff               0.9        1000            1  avgt    6      700.642 ±      12.765  ns/op
[info] BitSetBenchmark.oldDiff               0.9        1000         1000  avgt    6      741.179 ±      13.959  ns/op
[info] BitSetBenchmark.oldDiff               0.9        1000     10000000  avgt    6      770.903 ±      92.481  ns/op
[info] BitSetBenchmark.oldDiff               0.9    10000000            1  avgt    6  4340131.776 ± 1543854.072  ns/op
[info] BitSetBenchmark.oldDiff               0.9    10000000         1000  avgt    6  5059978.525 ±  545622.610  ns/op
[info] BitSetBenchmark.oldDiff               0.9    10000000     10000000  avgt    6  6337590.481 ± 9985141.843  ns/op
[info] BitSetBenchmark.oldDiff               1.0           1            1  avgt    6      423.524 ±       6.959  ns/op
[info] BitSetBenchmark.oldDiff               1.0           1         1000  avgt    6      430.071 ±       5.685  ns/op
[info] BitSetBenchmark.oldDiff               1.0           1     10000000  avgt    6      444.353 ±      38.419  ns/op
[info] BitSetBenchmark.oldDiff               1.0        1000            1  avgt    6      700.939 ±      19.550  ns/op
[info] BitSetBenchmark.oldDiff               1.0        1000         1000  avgt    6      765.482 ±      16.529  ns/op
[info] BitSetBenchmark.oldDiff               1.0        1000     10000000  avgt    6      754.800 ±      48.719  ns/op
[info] BitSetBenchmark.oldDiff               1.0    10000000            1  avgt    6  4355270.499 ± 1758132.946  ns/op
[info] BitSetBenchmark.oldDiff               1.0    10000000         1000  avgt    6  4854703.397 ± 2019329.042  ns/op
[info] BitSetBenchmark.oldDiff               1.0    10000000     10000000  avgt    6  4781573.211 ± 2910313.306  ns/op
```

## Benchmarks for filter
~ 50% - 70% improvement 
```
[info] Benchmark                  (sizeLeft)  Mode  Cnt          Score           Error  Units
[info] BitSetBenchmark.newFilter           1  avgt    6        105.774 ±        52.917  ns/op
[info] BitSetBenchmark.newFilter        1000  avgt    6       8796.433 ±      3451.048  ns/op
[info] BitSetBenchmark.newFilter    10000000  avgt    6  299675411.708 ±  13775556.597  ns/op
[info] BitSetBenchmark.oldFilter           1  avgt    6        620.988 ±       184.030  ns/op
[info] BitSetBenchmark.oldFilter        1000  avgt    6      37135.108 ±      5199.098  ns/op
[info] BitSetBenchmark.oldFilter    10000000  avgt    6  870425992.333 ± 363293650.619  ns/op
```

### CODE REVIEW

File: src/library/scala/collection/BitSet.scala [hunk 1]
```diff
@@ -227,6 +227,16 @@ trait BitSetOps[+C <: BitSet with BitSetOps[C]]
...
+  def oldDiff(other: Set[Int]): C = other match {
+    case otherBitset: BitSet =>
+      val len = coll.nwords
+      val words = new Array[Long](len)
+      for (idx <- 0 until len)
+        words(idx) = this.word(idx) & ~otherBitset.word(idx)
+      fromBitMaskNoCopy(words)
+    case _ => diff(other)
+  }
+
```
Comment (joshlemer) [inline src/library/scala/collection/BitSet.scala]: Left just for benchmarking...

**Comment by joshlemer** [comment]
## 💡 Ideas for future work:

the following methods can most likely be profitably optimized:
```
    sizeCompare()
    span()
    dropRight()
    drop()
    intersect()
    grouped()
    groupBy()
    splitAt()
    partition()
```

File: src/library/scala/collection/immutable/BitSet.scala [hunk 2]
```diff
@@ -39,6 +37,18 @@ sealed abstract class BitSet
...
+    val it = iterator
+    while (it.hasNext) {
+      val elem = it.next()
+      if (pred(elem) != isFlipped) {
+        b += elem
+      }
+    }
+    b.result()
+  }
+
```
Comment (joshlemer) [inline src/library/scala/collection/immutable/BitSet.scala]: Left for benchmarking, to be removed

File: src/library/scala/collection/immutable/BitSet.scala [hunk 3]
```diff
@@ -134,6 +132,21 @@ object BitSet extends SpecificIterableFactory[Int, BitSet] {
...
+        case 0 => this
+        case _ =>
+          val newElems = elems & ~bs.word(0)
+          if (newElems == 0L) empty else new BitSet1(newElems)
+      }
+      case _ => super.diff(other)
+    }
+
+    override def filterImpl(pred: Int => Boolean, isFlipped: Boolean): BitSet =
+      new BitSet1(BitSetOps.computeWordForFilter(pred, isFlipped, elems, 0))
```
Comment (lrytz) [inline src/library/scala/collection/immutable/BitSet.scala]: Would it make sense to check for `0L` and return `empty`?

File: src/library/scala/collection/immutable/BitSet.scala [hunk 4]
```diff
@@ -143,12 +156,151 @@ object BitSet extends SpecificIterableFactory[Int, BitSet] {
...
+        case _ => new BitSet2(elems0 & ~bs.word(0), elems1 & ~bs.word(1))
+      }
+      case _ => super.diff(other)
+    }
+
+    override def filterImpl(pred: Int => Boolean, isFlipped: Boolean): BitSet =
+      new BitSet2(
+        BitSetOps.computeWordForFilter(pred, isFlipped, elems0, 0),
+        BitSetOps.computeWordForFilter(pred, isFlipped, elems1, 1)
+      )
```
Comment (lrytz) [inline src/library/scala/collection/immutable/BitSet.scala]: Check if words are `0L`?

File: src/library/scala/collection/immutable/BitSet.scala [hunk 5]
```diff
@@ -143,12 +156,151 @@ object BitSet extends SpecificIterableFactory[Int, BitSet] {
       if (idx == 0) new BitSet2(w, elems1)
       else if (idx == 1) createSmall(elems0, w)
       else fromBitMaskNoCopy(updateArray(Array(elems0, elems1), idx, w))
+
+
+    override def diff(other: collection.Set[Int]): BitSet = other match {
+      case bs: collection.BitSet => bs.nwords match {
+        case 0 => this
+        case 1 => new BitSet2(elems0 & ~bs.word(0), elems1)
+        case _ => new BitSet2(elems0 & ~bs.word(0), elems1 & ~bs.word(1))
```
Comment (lrytz) [inline src/library/scala/collection/immutable/BitSet.scala]: check if words are `0L`?

File: src/library/scala/collection/mutable/BitSet.scala [hunk 6]
```diff
@@ -171,6 +172,95 @@ class BitSet(protected[collection] final var elems: Array[Long])
...
+        while (i >= 0) {
+          newElems(i) = word(i) & ~bs.word(i)
+          i -= 1
+        }
+        fromBitMaskNoCopy(newElems)
+      }
+    case _ => super.diff(that)
+  }
+
+  def oldFilter(pred: Int => Boolean, isFlipped: Boolean): BitSet = {
```
Comment (lrytz) [inline src/library/scala/collection/mutable/BitSet.scala]: remove this method?

**Review by lrytz — COMMENTED** [review]
In the immutable implementation, we could fall back to `BitSet1/2`, I pointed out a few spot, but also in `diff` / `filterImpl`. Did you consider this, would it make sense?

File: src/library/scala/collection/immutable/BitSet.scala (hunk 3 above)
Comment (joshlemer) [reply src/library/scala/collection/immutable/BitSet.scala]: I actually initially tried that but I didn't see improvement, in fact I saw that it performed worse. I should probably test again..

Comment (lrytz) [reply src/library/scala/collection/immutable/BitSet.scala]: Thanks! Even if it makes that operation (a little) slower, it's probably beneficial for future operations on the result.

**Review by lrytz — APPROVED** [review]
👍  thank you, @joshlemer

**Pull request merged**